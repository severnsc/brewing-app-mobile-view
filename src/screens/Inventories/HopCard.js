import React from "react";
import PropTypes from "prop-types";
import { Card, Subtitle, Text } from "../../components";

const HopCard = ({
  name,
  purchaseDate,
  quantity,
  alphaAcids,
  unitCost,
  countryOfOrigin
}) => (
  <Card
    upperLeft={<Subtitle value={name} />}
    upperRight={<Text value={purchaseDate} />}
    lowerLeft={<Text value={quantity} />}
    lowerCenterLeft={<Text value={alphaAcids} />}
    lowerCenterRight={<Text value={unitCost} />}
    lowerRight={<Text value={countryOfOrigin} />}
  />
);

HopCard.propTypes = {
  name: PropTypes.string,
  purchaseDate: PropTypes.string,
  quantity: PropTypes.string,
  alphaAcids: PropTypes.string,
  unitCost: PropTypes.string,
  countryOfOrigin: PropTypes.string
};

export default HopCard;
