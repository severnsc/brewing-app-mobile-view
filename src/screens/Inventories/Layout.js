import React from "react";
import PropTypes from "prop-types";
import { Tabs, Tab } from "../../components";
import { FlatList, View } from "react-native";

export const InventoriesLayout = ({
  onTabChange,
  data,
  renderItem,
  keyExtractor
}) => (
  <View>
    <Tabs onChange={onTabChange}>
      {(active, handlePress) => {
        const types = ["Hops", "Malt", "Yeast", "Other"];
        return (
          <React.Fragment>
            {types.map(type => (
              <Tab
                key={type}
                value={type}
                active={active === type}
                onPress={handlePress}
              />
            ))}
          </React.Fragment>
        );
      }}
    </Tabs>
    <FlatList data={data} renderItem={renderItem} keyExtractor={keyExtractor} />
  </View>
);

InventoriesLayout.propTypes = {
  onTabChange: PropTypes.func.isRequired,
  data: PropTypes.array,
  renderItem: PropTypes.func.isRequired,
  keyExtractor: PropTypes.func.isRequired
};

InventoriesLayout.defaultProps = {
  onTabChange: () => {},
  renderItem: () => null,
  keyExtractor: item => item.id
};

export default InventoriesLayout;
