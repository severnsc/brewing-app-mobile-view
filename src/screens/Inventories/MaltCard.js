import React from "react";
import PropTypes from "prop-types";
import { Card, Subtitle, Text, MaltColor } from "../../components";

const MaltCard = ({
  name,
  color,
  purchaseDate,
  quantity,
  maltType,
  unitCost,
  countryOfOrigin
}) => (
  <Card
    upperLeft={<Subtitle value={name} />}
    upperCenter={<MaltColor value={color} />}
    upperRight={<Text value={purchaseDate} />}
    lowerLeft={<Text value={quantity} />}
    lowerCenterLeft={<Text value={maltType} />}
    lowerCenterRight={<Text value={unitCost} />}
    lowerRight={<Text value={countryOfOrigin} />}
  />
);

MaltCard.propTypes = {
  name: PropTypes.string,
  color: PropTypes.number,
  purchaseDate: PropTypes.string,
  quantity: PropTypes.string,
  maltType: PropTypes.string,
  unitCost: PropTypes.string,
  countryOfOrigin: PropTypes.string
};

export default MaltCard;
