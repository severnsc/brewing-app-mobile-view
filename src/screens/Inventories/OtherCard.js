import React from "react";
import PropTypes from "prop-types";
import { Card, Subtitle, Text } from "../../components";

const OtherCard = ({ name, purchaseDate, quantity, unitCost }) => (
  <Card
    upperLeft={<Subtitle value={name} />}
    upperRight={<Text value={purchaseDate} />}
    lowerLeft={<Text value={quantity} />}
    lowerRight={<Text value={unitCost} />}
  />
);

OtherCard.propTypes = {
  name: PropTypes.string,
  purchaseDate: PropTypes.string,
  quantity: PropTypes.string,
  unitCost: PropTypes.string
};

export default OtherCard;
