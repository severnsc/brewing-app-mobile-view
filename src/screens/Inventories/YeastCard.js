import React from "react";
import PropTypes from "prop-types";
import { Card, Subtitle, Text } from "../../components";

const YeastCard = ({
  name,
  purchaseDate,
  quantity,
  yeastType,
  unitCost,
  dryOrLiquid
}) => (
  <Card
    upperLeft={<Subtitle value={name} />}
    upperRight={<Text value={purchaseDate} />}
    lowerLeft={<Text value={quantity} />}
    lowerCenterLeft={<Text value={yeastType} />}
    lowerCenterRight={<Text value={unitCost} />}
    lowerRight={<Text value={dryOrLiquid} />}
  />
);

YeastCard.propTypes = {
  name: PropTypes.string,
  purchaseDate: PropTypes.string,
  quantity: PropTypes.string,
  yeastType: PropTypes.string,
  unitCost: PropTypes.string,
  dryOrLiquid: PropTypes.oneOf(["Dry", "Liquid"])
};

export default YeastCard;
