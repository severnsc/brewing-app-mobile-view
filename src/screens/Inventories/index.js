import React, { useState } from "react";
import PropTypes from "prop-types";
import InventoriesLayout from "./Layout";
import HopCard from "./HopCard";
import MaltCard from "./MaltCard";
import YeastCard from "./YeastCard";
import OtherCard from "./OtherCard";

const Inventories = ({ data }) => {
  const [inventoryType, setInventoryType] = useState("Hops");
  const keyExtractor = item => item.id;
  const components = {
    Hops: item => (
      <HopCard
        name={item.name}
        purchaseDate={item.purchaseDate}
        quantity={item.quantity}
        alphaAcids={item.alphaAcids}
        unitCost={item.unitCost}
        countryOfOrigin={item.countryOfOrigin}
      />
    ),
    Malt: item => (
      <MaltCard
        name={item.name}
        color={item.color}
        purchaseDate={item.purchaseDate}
        quantity={item.quantity}
        maltType={item.maltType}
        unitCost={item.unitCost}
        countryOfOrigin={item.countryOfOrigin}
      />
    ),
    Yeast: item => (
      <YeastCard
        name={item.name}
        purchaseDate={item.purchaseDate}
        quantity={item.quantity}
        yeastType={item.yeastType}
        unitCost={item.unitCost}
        dryOrLiquid={item.dryOrLiquid}
      />
    ),
    Other: item => (
      <OtherCard
        name={item.name}
        purchaseDate={item.purchaseDate}
        quantity={item.quantity}
        unitCost={item.unitCost}
      />
    )
  };
  return (
    <InventoriesLayout
      onTabChange={setInventoryType}
      renderItem={components[inventoryType]}
      keyExtractor={keyExtractor}
      data={data}
    />
  );
};

Inventories.propTypes = {
  data: PropTypes.array
};

Inventories.defaultProps = {
  data: [{ id: "1" }]
};

export default Inventories;
