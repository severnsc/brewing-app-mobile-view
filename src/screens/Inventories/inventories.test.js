import React from "react";
import Inventories from ".";
import InventoriesLayout from "./Layout";
import { shallow } from "enzyme";

describe("Inventories layout", () => {
  it("returns tabs for each inventory category", () => {
    const inventories = shallow(<InventoriesLayout />);
    const tabs = inventories
      .find("Tabs")
      .props()
      .children("Hops", () => {});
    expect(tabs.props.children).toHaveLength(4);
    expect(tabs.props.children[0].props.value).toBe("Hops");
    expect(tabs.props.children[1].props.value).toBe("Malt");
    expect(tabs.props.children[2].props.value).toBe("Yeast");
    expect(tabs.props.children[3].props.value).toBe("Other");
  });
  it("passes onTabChange to Tabs", () => {
    const onTabChange = jest.fn();
    const inventories = shallow(
      <InventoriesLayout onTabChange={onTabChange} />
    );
    const tabs = inventories.find("Tabs");
    expect(tabs.prop("onChange")).toBe(onTabChange);
  });
  it("initially sets the Hops tab to active true", () => {
    const inventories = shallow(<InventoriesLayout />);
    const tabs = inventories.find("Tabs");
    const handlePress = jest.fn();
    const wrapper = tabs.props().children("Hops", handlePress);
    expect(wrapper.props.children[0].props.active).toBe(true);
    expect(wrapper.props.children[1].props.active).toBe(false);
    expect(wrapper.props.children[2].props.active).toBe(false);
    expect(wrapper.props.children[3].props.active).toBe(false);
  });
  it("sets the onPress of all tabs to handlePress", () => {
    const inventories = shallow(<InventoriesLayout />);
    const tabs = inventories.find("Tabs");
    const handlePress = jest.fn();
    const wrapper = tabs.props().children("Hops", handlePress);
    expect(wrapper.props.children[0].props.onPress).toBe(handlePress);
    expect(wrapper.props.children[1].props.onPress).toBe(handlePress);
    expect(wrapper.props.children[2].props.onPress).toBe(handlePress);
    expect(wrapper.props.children[3].props.onPress).toBe(handlePress);
  });
  it("returns a FlatList with renderItem from the data", () => {
    const data = [
      {
        id: "1",
        name: "Chinook",
        createdAt: "11/15/2018",
        quantity: 8.5,
        alphaAcids: 12.4,
        unitCost: 12,
        country: "USA"
      },
      {
        id: "2",
        name: "Bravo",
        createdAt: "12/2/2018",
        quantity: 10,
        alphaAcids: 10.6,
        unitCost: 9,
        country: "USA"
      }
    ];
    const renderItem = item => item;
    const inventories = shallow(
      <InventoriesLayout data={data} renderItem={renderItem} />
    );
    const list = inventories.find("FlatList");
    expect(list).toHaveLength(1);
    const cards = data.map(list.props().renderItem);
    expect(cards).toMatchSnapshot();
  });
});

describe("Inventories screen", () => {
  it("returns Hop cards when hops tab is active", () => {
    const data = [
      {
        id: "1",
        name: "Chinook",
        purchaseDate: "11/15/2018",
        quantity: "8.5 lbs",
        alphaAcids: "12.4% AA",
        unitCost: "$12/lb",
        countryOfOrigin: "USA"
      },
      {
        id: "2",
        name: "Bravo",
        purchaseDate: "12/2/2018",
        quantity: "10 lbs",
        alphaAcids: "10.6% AA",
        unitCost: "$9/lb",
        countryOfOrigin: "USA"
      }
    ];
    const inventories = shallow(<Inventories data={data} />);
    const layout = inventories.dive();
    const list = layout.find("FlatList");
    const cards = data.map(list.props().renderItem);
    expect(cards).toMatchSnapshot();
  });
  it("returns Malt cards when malt is active", () => {
    const data = [
      {
        id: "1",
        name: "Maris Otter",
        purchaseDate: "11/15/2018",
        quantity: "8.5 lbs",
        maltType: "Base",
        unitCost: "$3/lb",
        countryOfOrigin: "UK"
      },
      {
        id: "2",
        name: "Crystal 10",
        purchaseDate: "1/16/2019",
        quantity: "2.5 lbs",
        maltType: "Crystal",
        unitCost: "$3.50/lb",
        countryOfOrigin: "UK"
      }
    ];
    const inventories = shallow(<Inventories data={data} />);
    inventories.props().onTabChange("Malt");
    inventories.update();
    const layout = inventories.dive();
    const list = layout.find("FlatList");
    const cards = data.map(list.props().renderItem);
    expect(cards).toMatchSnapshot();
  });
  it("returns Yeast cards when yeast is active", () => {
    const data = [
      {
        id: "1",
        name: "WLP 001 - California Ale",
        purchaseDate: "11/15/2018",
        quantity: "1 vial",
        yeastType: "Ale",
        unitCost: "$8/vial",
        dryOrLiquid: "Liquid"
      },
      {
        id: "2",
        name: "WLP 002 - English Ale",
        purchaseDate: "1/15/2019",
        quantity: "2 vials",
        yeastType: "Ale",
        unitCost: "$10/vial",
        dryOrLiquid: "Liquid"
      }
    ];
    const inventories = shallow(<Inventories data={data} />);
    inventories.props().onTabChange("Yeast");
    inventories.update();
    const layout = inventories.dive();
    const list = layout.find("FlatList");
    const cards = data.map(list.props().renderItem);
    expect(cards).toMatchSnapshot();
  });
  it("returns Other cards when other is active", () => {
    const data = [
      {
        id: "1",
        name: "Bourbon",
        purchaseDate: "11/15/2018",
        quantity: "1 bottle",
        unitCost: "$25/bottle"
      },
      {
        id: "2",
        name: "Cocoa nibs",
        purchaseDate: "1/15/2019",
        quantity: "2 lbs",
        unitCost: "$10/lb"
      }
    ];
    const inventories = shallow(<Inventories data={data} />);
    inventories.props().onTabChange("Other");
    inventories.update();
    const layout = inventories.dive();
    const list = layout.find("FlatList");
    const cards = data.map(list.props().renderItem);
    expect(cards).toMatchSnapshot();
  });
});
