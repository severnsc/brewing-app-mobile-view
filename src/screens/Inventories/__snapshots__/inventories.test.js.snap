// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inventories layout returns a FlatList with renderItem from the data 1`] = `
Array [
  Object {
    "alphaAcids": 12.4,
    "country": "USA",
    "createdAt": "11/15/2018",
    "id": "1",
    "name": "Chinook",
    "quantity": 8.5,
    "unitCost": 12,
  },
  Object {
    "alphaAcids": 10.6,
    "country": "USA",
    "createdAt": "12/2/2018",
    "id": "2",
    "name": "Bravo",
    "quantity": 10,
    "unitCost": 9,
  },
]
`;

exports[`Inventories screen returns Hop cards when hops tab is active 1`] = `
Array [
  <HopCard
    alphaAcids="12.4% AA"
    countryOfOrigin="USA"
    name="Chinook"
    purchaseDate="11/15/2018"
    quantity="8.5 lbs"
    unitCost="$12/lb"
  />,
  <HopCard
    alphaAcids="10.6% AA"
    countryOfOrigin="USA"
    name="Bravo"
    purchaseDate="12/2/2018"
    quantity="10 lbs"
    unitCost="$9/lb"
  />,
]
`;

exports[`Inventories screen returns Malt cards when malt is active 1`] = `
Array [
  <MaltCard
    countryOfOrigin="UK"
    maltType="Base"
    name="Maris Otter"
    purchaseDate="11/15/2018"
    quantity="8.5 lbs"
    unitCost="$3/lb"
  />,
  <MaltCard
    countryOfOrigin="UK"
    maltType="Crystal"
    name="Crystal 10"
    purchaseDate="1/16/2019"
    quantity="2.5 lbs"
    unitCost="$3.50/lb"
  />,
]
`;

exports[`Inventories screen returns Other cards when other is active 1`] = `
Array [
  <OtherCard
    name="Bourbon"
    purchaseDate="11/15/2018"
    quantity="1 bottle"
    unitCost="$25/bottle"
  />,
  <OtherCard
    name="Cocoa nibs"
    purchaseDate="1/15/2019"
    quantity="2 lbs"
    unitCost="$10/lb"
  />,
]
`;

exports[`Inventories screen returns Yeast cards when yeast is active 1`] = `
Array [
  <YeastCard
    dryOrLiquid="Liquid"
    name="WLP 001 - California Ale"
    purchaseDate="11/15/2018"
    quantity="1 vial"
    unitCost="$8/vial"
    yeastType="Ale"
  />,
  <YeastCard
    dryOrLiquid="Liquid"
    name="WLP 002 - English Ale"
    purchaseDate="1/15/2019"
    quantity="2 vials"
    unitCost="$10/vial"
    yeastType="Ale"
  />,
]
`;
